Bluetooth: AVRCP: Implement unit message reception on the AVRCP target

makeshi opened the pull request:
@gzh-terry Please help review it, thanks.

gzh-terry commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -177,6 +177,13 @@ struct bt_avrcp_cb {
 	 *  @param rsp The response for PASS THROUGH command.
 	 */
 	void (*passthrough_rsp)(struct bt_avrcp *avrcp, struct bt_avrcp_passthrough_rsp *rsp);
+	/** @brief Callback function called when received unit info command
```
Leave one line between 179 and 180 would be better :)

makeshi replied:
Thanks， I remember there was no space before in these call back function list, and I'd like to keep it consistent. Anyway, I'll add a space.

gzh-terry replied:
> Thanks， I remember there was no space before in these call back function list, and I'd like to keep it consistent. :) Anyway, I'll add a space.

Yes, I added it here [#84367](https://github.com/zephyrproject-rtos/zephyr/pull/84367) :)

gzh-terry commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -177,6 +177,13 @@ struct bt_avrcp_cb {
 	void (*passthrough_rsp)(struct bt_avrcp *avrcp, struct bt_avrcp_passthrough_rsp *rsp);
+	/** @brief Callback function called when received unit info command
+	 *
+	 *  This callback notifies the application that a avrcp unit info command is received.
```
"a avrcp" -> "an AVRCP"

gzh-terry commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -177,6 +177,13 @@ struct bt_avrcp_cb {
+	 *
+	 *  @param avrcp AVRCP connection object.
+	 */
+	void (*unit_info_cmd_recv)(struct bt_avrcp *avrcp);
```
I would recommend `(*unit_info_req)` to keep the style consistent with other AVRCP (and A2DP as well) callbacks.

gzh-terry commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -249,6 +256,17 @@ int bt_avrcp_get_subunit_info(struct bt_avrcp *avrcp);
 int bt_avrcp_passthrough(struct bt_avrcp *avrcp, bt_avrcp_opid_t operation_id,
 			 bt_avrcp_button_state_t state, const uint8_t *payload, uint8_t len);
 
+/** @brief Response unit info command.
```
Response -> Respond

makeshi replied:
Sure, will update.

gzh-terry commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -249,6 +256,17 @@ int bt_avrcp_get_subunit_info(struct bt_avrcp *avrcp);
 
+/** @brief Response unit info command.
+ *
+ *  Send the response unit info command when the receiver will be notified by the registered
```
response unit info command -> unit info response

gzh-terry commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -249,6 +256,17 @@ int bt_avrcp_get_subunit_info(struct bt_avrcp *avrcp);
+/** @brief Response unit info command.
+ *
+ *  Send the response unit info command when the receiver will be notified by the registered
+ *  callback (unit_info_cmd_recv).
```
Typically we do not assume the peer device uses the same Bluetooth stack. I would recommend remove "when the receiver will ... callback (unit_info_cmd_recv)"

makeshi replied:
Actually, there is no relation between peer device,  the function and callback should only run the same device and also in target mode .But to avoid  such confusion, I will remove this description.

gzh-terry replied:
Sorry for misundersdanding, probobaly you want say "Send the unit info response when notified by the registered callback (unit_info_cmd_recv)."

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -649,3 +666,27 @@ int bt_avrcp_register_cb(const struct bt_avrcp_cb *cb)
+int bt_avrcp_response_unit_info(struct bt_avrcp *avrcp, struct bt_avrcp_unit_info_rsp *rsp)
+{
+	struct net_buf *buf;
+	uint8_t data[5];
```
Try to keep style consistent, e.g., use param[5].

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -649,3 +666,27 @@ int bt_avrcp_register_cb(const struct bt_avrcp_cb *cb)
+	/* Add additional info */
+	data[i++] = 0x07;
+	/* Add Unit Type info */
+	data[i++] = (rsp->unit_type) << 3;
```
Use `FIELD_PREP()` to avoid bit order issue.

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -649,3 +666,27 @@ int bt_avrcp_register_cb(const struct bt_avrcp_cb *cb)
+	/* Add Unit Type info */
+	data[i++] = (rsp->unit_type) << 3;
+	/* Company ID */
+	data[i++] = (uint8_t)((rsp->company_id) >> 16);
```
Similarly, I would recomment `net_buf_add_be24`

makeshi replied:
will use net_buf function to fill data

gzh-terry commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -49,6 +49,24 @@ static void avrcp_unit_info_rsp(struct bt_avrcp *avrcp, struct bt_avrcp_unit_inf
+	rsp.unit_type = 0x9;
+	rsp.company_id = 0x001958; /*BT SIG registered Company ID*/
+
+	err = bt_avrcp_response_unit_info(avrcp, &rsp);
```
Caution: this is in the callback thread, and might lead to assertion in net_buf.

makeshi replied:
Thanks, let me check it

gzh-terry commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -49,6 +49,24 @@ static void avrcp_unit_info_rsp(struct bt_avrcp *avrcp, struct bt_avrcp_unit_inf
+	if (!err) {
+		bt_shell_print("AVRCP target response unit info");
+	} else {
+		bt_shell_print("failed to response unit info");
```
respond?

gzh-terry commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -249,6 +257,16 @@ int bt_avrcp_get_subunit_info(struct bt_avrcp *avrcp);
 int bt_avrcp_passthrough(struct bt_avrcp *avrcp, bt_avrcp_opid_t operation_id,
 			 bt_avrcp_button_state_t state, const uint8_t *payload, uint8_t len);
```
Set -> Send?

makeshi replied:
It's also okay to use "send". I use set because I want it to correspond to the above API bt_avrcp_get_unit_info.

gzh-terry replied:
> It's also okay to use "send". I use set because I want it to correspond to the above API bt_avrcp_get_unit_info.

I think `set` usualy means to provide a value and save at the local device, or set a configuration to the remote.
However, this interface is to inform the remote device, rather than a configuration.

makeshi replied:
Agree, Here using "Send" here can indeed better express the actual meaning than "Set" for this function. I'll update the comments and the function name.

gzh-terry reviewed (dismissed):
Generally fine to me.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -253,9 +253,26 @@ static void avrcp_unit_info_handler(struct bt_avrcp *avrcp, struct net_buf *buf,
-		/* ToDo */
+		if ((avrcp_cb != NULL) && (avrcp_cb->unit_info_req != NULL)) {
+			avrcp_hdr = (void *)buf->data;
+			net_buf_pull(buf, sizeof(*avrcp_hdr));
```
The `buf->len` needs to be checked before pulling the net buffer.

After checking the whole receiving data path, I found an issue that the `buf->len` is also not checked before accessing the `buf->data` in function `avrcp_recv`.
https://github.com/zephyrproject-rtos/zephyr/blob/7ee4011166ae0467d6830bccd337f9c8f583acda/subsys/bluetooth/host/classic/avrcp.c#L354-L366

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -253,9 +253,26 @@ static void avrcp_unit_info_handler(struct bt_avrcp *avrcp, struct net_buf *buf,
 	struct bt_avrcp_unit_info_rsp rsp;
+	bt_avrcp_subunit_type_t subunit_type;
 
 	if (cr == BT_AVCTP_CMD) {
```
I think it is better to split the handler for request and response.

makeshi replied:
I will split the AVRCP code to handle with different role CT and CG.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -253,9 +253,26 @@ static void avrcp_unit_info_handler(struct bt_avrcp *avrcp, struct net_buf *buf,
+			}
+
+			subunit_type = BT_AVRCP_HDR_GET_SUBUNIT_TYPE(avrcp_hdr);
+			if ((subunit_type != BT_AVRCP_SUBUNIT_TYPE_UNIT) ||
+				(avrcp_hdr->opcode != BT_AVRCP_OPC_UNIT_INFO)) {
+				LOG_ERR("Invalid unit info command");
+				return;
+			}
```
For the command, if the command is invalid, the response should be sent internally instead of missing it.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -253,9 +253,26 @@ static void avrcp_unit_info_handler(struct bt_avrcp *avrcp, struct net_buf *buf,
+			if ((subunit_type != BT_AVRCP_SUBUNIT_TYPE_UNIT) ||
+				(avrcp_hdr->opcode != BT_AVRCP_OPC_UNIT_INFO)) {
+				LOG_ERR("Invalid unit info command");
+				return;
```
For the command, if the command is invalid, the response should be sent internally instead of missing it.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -253,9 +253,26 @@ static void avrcp_unit_info_handler(struct bt_avrcp *avrcp, struct net_buf *buf,
 
 	if (cr == BT_AVCTP_CMD) {
-		/* ToDo */
+		if ((avrcp_cb != NULL) && (avrcp_cb->unit_info_req != NULL)) {
```
The command handling should be controlled by the registered callback. Such as, if the command is not valid or the command is unsupported, the error response should be sent directly.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -649,3 +666,22 @@ int bt_avrcp_register_cb(const struct bt_avrcp_cb *cb)
+{
+	struct net_buf *buf;
+
+	buf = avrcp_create_unit_pdu(avrcp, BT_AVCTP_RESPONSE);
```
The `TID` should be same with the `TID` of the command.

makeshi replied:
TID should added as  all function calls and callback parameter, handled in application code side.

lylezhu2012 replied:
But the TID cannot be set by application in your PR.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -649,3 +666,22 @@ int bt_avrcp_register_cb(const struct bt_avrcp_cb *cb)
+	}
+
+	/* Add additional info */
+	net_buf_add_u8(buf, 0x07);
```
What does the mean of `7`? I think the hard code should be defined as a MICRO.

makeshi replied:
I did not find the real mean of the 7, 
![image](https://github.com/user-attachments/assets/d566e98d-6fe2-4397-bd33-8ebf51b9cb2e)
and The protocol analysis tool recognizes as the "Const byte"
![image](https://github.com/user-attachments/assets/10fc7f94-802c-4277-a01e-e2b8032eacf1)

gzh-terry replied:
Hi @makeshi  @lylezhu2012 

`0x07` means `page = 0` and `extension_code = 7`
`page` shall only be greater than 0 when the `number of extended units` exceeds 5, which is not possible at unit info response, where the `number of extended units` is 0.
`extension_code` is not defined currently, so all three bits are set to 1.

![image](https://github.com/user-attachments/assets/72f80514-a0e4-438e-898f-55455d6f2c4c)

gzh-terry replied:
UNIT INFO response generally has the same structure as SUBUNIT INFO response.

gzh-terry replied:
@makeshi  can use `AVRCP_SUBUNIT_EXTENSION_COED`

lylezhu2012 replied:
> Hi @makeshi @lylezhu2012
> 
> `0x07` means `page = 0` and `extension_code = 7` `page` shall only be greater than 0 when the `number of extended units` exceeds 4, which is not possible at unit info response, where the `number of extended units` is 0. `extension_code` is not defined currently, so all three bits are set to 1.

The command is `unit info` instead of `subunit info`. 
From `UNIT INFO status command response format`, it is a hardcode without any comments. It seams the field has not yet explained of what it means.
![image](https://github.com/user-attachments/assets/a29b1984-93b4-48c4-a29b-849990c86fa0)

gzh-terry replied:
@makeshi 
BTW, found a typo here, `COED` -> `CODE`
Can you help me to correct it if using this macro?

gzh-terry replied:
> > Hi @makeshi @lylezhu2012
> > `0x07` means `page = 0` and `extension_code = 7` `page` shall only be greater than 0 when the `number of extended units` exceeds 4, which is not possible at unit info response, where the `number of extended units` is 0. `extension_code` is not defined currently, so all three bits are set to 1.
> 
> The command is `unit info` instead of `subunit info`. From `UNIT INFO status command response format`, it is a hardcode without any comments. It seams the field has not yet explained of what it means. ![image](https://private-user-images.githubusercontent.com/47838210/416062789-a29b1984-93b4-48c4-a29b-849990c86fa0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDAzNzA5MTksIm5iZiI6MTc0MDM3MDYxOSwicGF0aCI6Ii80NzgzODIxMC80MTYwNjI3ODktYTI5YjE5ODQtOTNiNC00OGM0LWEyOWItODQ5OTkwYzg2ZmEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI0VDA0MTY1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA2NmJlOTcwYTk5Y2NjNWYyZDliMjFjY2M2NjQxZDdlOGI0MDFiNmE0MTYzNTg1NDRhZjYxMzY5OGM0YTgyN2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.UEhZaplDruN6HhytqMpN6MnhmgRvoD13i8Z5DWr9wvo)

Then just put `0x7` is fine, to exactly match the spec.

jhedberg replied:
It's good to have at least a code comment pointing out that the constant is hard-coded in the spec

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -649,3 +666,22 @@ int bt_avrcp_register_cb(const struct bt_avrcp_cb *cb)
+	/* Add additional info */
+	net_buf_add_u8(buf, 0x07);
+	/* Add Unit Type info */
+	net_buf_add_u8(buf, FIELD_PREP(GENMASK(7, 3), (rsp->unit_type)));
```
For `AVRCP`, the `unit_type` should be `Panel`, right?

makeshi replied:
current , it is right, but in spec, it mentions
![image](https://github.com/user-attachments/assets/1429ed5a-a260-4d3a-9d2a-87c9a13d92d1)
So, for the code extensibility, I also used the uint_type as a parameter not hard code for Panel

gzh-terry replied:
@makeshi @lylezhu2012 
We may consider to expose `bt_avrcp_subunit_type_t` (`BT_AVRCP_SUBUNIT_TYPE_PANEL = 0x9`) to users?

makeshi replied:
Yes, I think we could expose it

nashif commented:
Please set a readable and useful PR title and add more description the PR body, the PR body should be used for more details about the change and not to ask for reviews.

jhedberg commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -243,13 +266,14 @@ static void avrcp_disconnected(struct bt_avctp *session)
 	}
 }
 
-static void avrcp_vendor_dependent_handler(struct bt_avrcp *avrcp, struct net_buf *buf,
+static void avrcp_ct_vendor_dependent_handler(struct bt_avrcp *avrcp, struct net_buf *buf,
 					   bt_avctp_cr_t cr)
```
While you're doing these changes, could you also fix the alignment of the fixed lines? (everywhere - not just here).

makeshi replied:
OK, thanks, I will fix them.

jhedberg commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -214,11 +223,23 @@ static struct bt_avrcp *get_new_connection(struct bt_conn *conn)
+	struct bt_avrcp *avrcp = AVRCP_KCTWORK(work);
 
-	LOG_WRN("Timeout: tid 0x%X, opc 0x%02X", avrcp->req.tid, avrcp->req.opcode);
+	LOG_WRN("Controller timeout: tid 0x%X, opc 0x%02X", avrcp->ct_req.tid,
+		 avrcp->ct_req.opcode);
```
minor: should be aligned with the `(` and not the `"`

makeshi replied:
Thanks, updated now.

jhedberg commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -37,6 +37,12 @@ typedef enum __packed {
 
+/** @brief AV/C subunit type, also used for unit type */
+typedef enum __packed {
+	BT_AVRCP_SUBUNIT_TYPE_PANEL = 0x9,
+	BT_AVRCP_SUBUNIT_TYPE_UNIT = 0x1F,
```
Please be consistent with formatting. Writing `0x9` implies that the size is 4 bits, but that's clearly not true when the next value is `0x1F` (which should be `0x1f` for consistency, just look at your OPID definitions below).

makeshi replied:
Good catch, will update it.

jhedberg commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -377,20 +471,39 @@ static int avrcp_recv(struct bt_avctp *session, struct net_buf *buf)
+		avrcp->tg_req.subunit = subunit_type;
+		avrcp->tg_req.opcode = avrcp_hdr->opcode;
+
+		for (i = 0U; i < ARRAY_SIZE(tg_handler); i++) {
```
You can use `ARRAY_FOR_EACH(tg_handler, i) {`

jhedberg commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -567,10 +681,27 @@ static int avrcp_send(struct bt_avrcp *avrcp, struct net_buf *buf)
+			/* No max response time for pass through commands  */
+		} else if (tid != avrcp->tg_req.tid || subunit_type != avrcp->tg_req.subunit ||
+			   avrcp_hdr->opcode != avrcp->tg_req.opcode) {
+			LOG_WRN("unexpected AVRCP response, expected tid:0x%X, subunit:0x%X, "
+				"opc:0x%02X",
```
The fact that you have so many levels of indentation that you're forced to split up this string literal is a good indication that this `avrcp_send()` function could use some refactoring as well.

makeshi replied:
OK, thanks, will update it.

makeshi replied:
Updated in the latest code.

jhedberg commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -579,22 +710,34 @@ static int avrcp_send(struct bt_avrcp *avrcp, struct net_buf *buf)
+	}
+
+	return avrcp_send(avrcp, buf);
+}
 int bt_avrcp_get_unit_info(struct bt_avrcp *avrcp)
```
Missing empty line after `}`

makeshi replied:
updated in the latest code.

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -332,11 +356,77 @@ static void avrcp_pass_through_handler(struct bt_avrcp *avrcp, struct net_buf *b
+	bt_avrcp_subunit_type_t subunit_type;
+	int err;
+
+	while (1) {
+		if (cr == BT_AVCTP_CMD) {
+			if ((avrcp_cb != NULL) && (avrcp_cb->unit_info_req != NULL)) {
```
Thanks, good suggestion, will update it.

makeshi replied:
I the latest code, CR already have been removed for handle code.

makeshi commented:
Based on the current understanding of the TG unit integration from current implementation for unit message reception, and combined with your review comments, I now believe we need to formally separate CT and TG not only in their implementations, but also at the API interface and data structure level. Specifically:
1. CT and TG should each have independent callback functions and structure name
2. They should have separate connect APIs
3. All APIs should be distinguished using _ct/_tg suffixes to indicate ownership
Here is use pseudocode to explain this:
CT:
struct bt_avrcp_ct_cb {
	void (*connected)(struct bt_avrcp_ct *avrcp);
	void (*disconnected)(struct bt_avrcp_ct *avrcp);
         ......
}
int bt_avrcp_ct_register_cb(const struct bt_avrcp_ct_cb *cb);
struct bt_avrcp_ct *bt_avrcp_ct_connect(struct bt_conn *conn);
int bt_avrcp_ct_disconnect(struct bt_avrcp_ct *avrcp);
......
TG:
struct bt_avrcp_tg_cb {
	void (*connected)(struct bt_avrcp_tg *avrcp);
	void (*disconnected)(struct bt_avrcp_tg *avrcp);
         ......
}
int bt_avrcp_tg_register_cb(const struct bt_avrcp_tg_cb *cb);
struct bt_avrcp_tg *bt_avrcp_tg_connect(struct bt_conn *conn);
int bt_avrcp_tg_disconnect(struct bt_avrcp_tg *avrcp);
......


@gzh-terry @lylezhu2012 @jhedberg If you also agree, I would like to submit a PR to rename the current CT-related APIs (adding _ct suffixes) before I finalize and submit my TG unit implementation PR.

jhedberg commented:
@makeshi doing the CT/TG separation sounds correct to me

gzh-terry commented:
> Based on the current understanding of the TG unit integration from current implementation for unit message reception, and combined with your review comments, I now believe we need to formally separate CT and TG not only in their implementations, but also at the API interface and data structure level. Specifically:
> 
> 1. CT and TG should each have independent callback functions and structure name
> 2. They should have separate connect APIs
> 3. All APIs should be distinguished using _ct/_tg suffixes to indicate ownership
>    Here is use pseudocode to explain this:
>    CT:
>    struct bt_avrcp_ct_cb {
>    void (*connected)(struct bt_avrcp_ct *avrcp);
>    void (*disconnected)(struct bt_avrcp_ct *avrcp);
>    ......
>    }
>    int bt_avrcp_ct_register_cb(const struct bt_avrcp_ct_cb *cb);
>    struct bt_avrcp_ct *bt_avrcp_ct_connect(struct bt_conn *conn);
>    int bt_avrcp_ct_disconnect(struct bt_avrcp_ct *avrcp);
>    ......
>    TG:
>    struct bt_avrcp_tg_cb {
>    void (*connected)(struct bt_avrcp_tg *avrcp);
>    void (*disconnected)(struct bt_avrcp_tg *avrcp);
>    ......
>    }
>    int bt_avrcp_tg_register_cb(const struct bt_avrcp_tg_cb *cb);
>    struct bt_avrcp_tg *bt_avrcp_tg_connect(struct bt_conn *conn);
>    int bt_avrcp_tg_disconnect(struct bt_avrcp_tg *avrcp);
>    ......
> 
> @gzh-terry @lylezhu2012 @jhedberg If you also agree, I would like to submit a PR to rename the current CT-related APIs (adding _ct suffixes) before I finalize and submit my TG unit implementation PR.

Hi @makeshi,

However, separating them into two groups may cause ambiguity on the app side (same issue in Android). Indeed, CT and TG share the same L2CAP connection.
Specifically, for a device that supports absolute volume, it would act as CT and TG simultaneously. Then, the app must call both `bt_avrcp_ct_connect` and `bt_avrcp_tg_connect`, these lead to a conflict at AVCTP level.
The second issue that would soon arise is when the app calls `bt_avrcp_ct_connect`, we cannot know if the TG service should be added in SDP.

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -33,8 +33,13 @@ LOG_MODULE_REGISTER(bt_avrcp);
+	struct bt_avrcp_req ct_req;
+	struct bt_avrcp_req tg_req;
+	struct k_work_delayable ct_timeout_work;
+	struct k_work_delayable tg_timeout_work;
```
Hi @makeshi 

As discussed with @lylezhu2012 last week, we might remove all TID and timeout-related codes at the stack level and move these operations to the app. The reasons are:

1. There is nothing done at the stack side but a print.
2. The timeout mechanism makes the logic too complicated and is rarely used (Based on our experience, no product is using this timeout, and the application typically has its error handling by other methods).
3. As described in AVCTP v1.4: "On the controller and the target side, handling of transaction labels is dependent on the application."

makeshi replied:
yes, I am fine when we move the TID logic to application side, I also will remove it.

gzh-terry commented:
Actually, in the majority of cases, both CT and TG are supported (due to the absolute volume feature). Then how do we generate a callback when AVRCP is connected? Is it one of the two `connected`, or both?

gzh-terry commented:
And this is what Android did: [Reference](https://cs.android.com/android/platform/superproject/main/+/main:packages/modules/Bluetooth/system/include/hardware/bt_rc.h;drc=d7acd7d5623f67aa959ae23a362c337288e96d69;l=374?hl=zh-cn)

It is assumed that only one of CT and TG is selected, and the role of absolute volume related commands is reversed (i.e., TG sends a command, and CT sends a response).

makeshi commented:
> (due to the absolute volume



> both

For my understanding, it depends on whether the CT or TG is selected. If both are selected when  bt_avrcp_ct_register_cb  and  bt_avrcp_tg_register_cb  are called, both connection callbacks should be called , and they should return both the AVRCP CT and TG handles. Then, the applications can use these CT and TG handles to do operations for the different roles of AVRCP.

makeshi commented:
Disscussed with lyle offline, proposal as bellow:
1. Only the connect and disconnect APIs of AVRCP are made AVRCP public api for CT and TG. 
2. The connect API no longer returns the AVRCP handle. The AVRCP handle is returned by the connected callback of the CT or TG . The connected API also returns the ACL connection and AVRCP role handle at the same time.
3. And all the other callback functions and feature APIs separated and all feature APIs have the prefixes _ct or _tg.
The bt_avrcp_ct_register_cb or bt_avrcp_tg_register_cb is called can be used to confirm which role is supported.
Here is use pseudocode to explain this:

int bt_avrcp_connect(struct bt_conn *conn);
int bt_avrcp_disconnect(struct bt_conn *conn);
 
struct bt_avrcp_ct_cb {
	void (*connected)(struct bt_conn *conn, struct bt_avrcp_ct *ct);
	void (*disconnected)(struct bt_avrcp_ct *ct);
	.......
};
int bt_avrcp_ct_register_cb(const struct bt_avrcp_ct_cb *cb);
int bt_avrcp_ct_get_unit_info(struct bt_avrcp_ct *ct, uint8_t tid);
 
struct bt_avrcp_tg_cb {
	void (*connected)(struct bt_conn *conn, struct bt_avrcp_tg *tg);
	void (*disconnected)(struct bt_avrcp_tg *tg);
	void (*unit_info_req)(struct bt_avrcp_tg *tg, uint8_t tid);
	.......
};
int bt_avrcp_tg_register_cb(const struct bt_avrcp_tg_cb *cb);
int bt_avrcp_tg_send_unit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid, struct bt_avrcp_unit_info_rsp *rsp);

The shell demo commands are also divided into those related to AVRCP change. The AVRCP commands are only responsible
for connect and disconnect, and the sub-functions commands under AVRCP are used to  avrcp ct and avrcp tg. 
 
uart:~$ avrcp
avrcp - Bluetooth AVRCP sh commands
Subcommands:
  connect     : connect AVRCP
  disconnect  : disconnect AVRCP
  ct          : AVRCP CT shell commands
  tg          : AVRCP TG shell commands
uart:~$ avrcp ct
ct - AVRCP CT shell commands
Subcommands:
  register_ct_cb  : register avrcp callbacks
  get_unit        : get unit info
  get_subunit     : get subunit info
  get_cap         : get capabilities <cap_id: company or events>
  play            : request a play at the remote player
  pause           : request a pause at the remote player
uart:~$ avrcp tg
tg - AVRCP TG shell commands
Subcommands:
  register_cb    : register avrcp callbacks
  send_unit_rsp  : send unit info response
uart:~$

Does the AVRCP api refactor is accept to you? please comments if you have any concern.

makeshi commented:
> And this is what Android did: [Reference](https://cs.android.com/android/platform/superproject/main/+/main:packages/modules/Bluetooth/system/include/hardware/bt_rc.h;drc=d7acd7d5623f67aa959ae23a362c337288e96d69;l=374?hl=zh-cn)
> 
> It is assumed that only one of CT and TG is selected, and the role of absolute volume related commands is reversed (i.e., TG sends a command, and CT sends a response).
If my understanding is right, if in the new design case in zephyr avrcp, 
user should select the two role by register ct/tg callback both, suppose no such the role of absolute volume related commands is reversed illusion.

gzh-terry commented:
> Discussed with lyle offline, proposal as bellow:
> 
> 1. Only the connect and disconnect APIs of AVRCP are made AVRCP public api for CT and TG.
> 2. The connect API no longer returns the AVRCP handle. The AVRCP handle is returned by the connected callback of the CT or TG . The connected API also returns the ACL connection and AVRCP role handle at the same time.
> 3. And all the other callback functions and feature APIs separated and all feature APIs have the prefixes _ct or _tg.
>    The bt_avrcp_ct_register_cb or bt_avrcp_tg_register_cb is called can be used to confirm which role is supported.
>    Here is use pseudocode to explain this:
> 
> int bt_avrcp_connect(struct bt_conn *conn); int bt_avrcp_disconnect(struct bt_conn *conn);
> 
> struct bt_avrcp_ct_cb { void (*connected)(struct bt_conn *conn, struct bt_avrcp_ct *ct); void (*disconnected)(struct bt_avrcp_ct *ct); ....... }; int bt_avrcp_ct_register_cb(const struct bt_avrcp_ct_cb *cb); int bt_avrcp_ct_get_unit_info(struct bt_avrcp_ct *ct, uint8_t tid);
> 
> struct bt_avrcp_tg_cb { void (*connected)(struct bt_conn *conn, struct bt_avrcp_tg *tg); void (*disconnected)(struct bt_avrcp_tg *tg); void (*unit_info_req)(struct bt_avrcp_tg *tg, uint8_t tid); ....... }; int bt_avrcp_tg_register_cb(const struct bt_avrcp_tg_cb *cb); int bt_avrcp_tg_send_unit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid, struct bt_avrcp_unit_info_rsp *rsp);
> 
> The shell demo commands are also divided into those related to AVRCP change. The AVRCP commands are only responsible for connect and disconnect, and the sub-functions commands under AVRCP are used to avrcp ct and avrcp tg.
> 
> uart:~$ avrcp avrcp - Bluetooth AVRCP sh commands Subcommands: connect : connect AVRCP disconnect : disconnect AVRCP ct : AVRCP CT shell commands tg : AVRCP TG shell commands uart:~$ avrcp ct ct - AVRCP CT shell commands Subcommands: register_ct_cb : register avrcp callbacks get_unit : get unit info get_subunit : get subunit info get_cap : get capabilities <cap_id: company or events> play : request a play at the remote player pause : request a pause at the remote player uart:~$ avrcp tg tg - AVRCP TG shell commands Subcommands: register_cb : register avrcp callbacks send_unit_rsp : send unit info response uart:~$
> 
> @gzh-terry Does the AVRCP api refactor is accept to you? please comments if you have any concern.

Look good to me.

makeshi committed:
Bluetooth: AVRCP: Refactored AVRCP API for CT/TG separation

- Only AVRCP connect and disconnect api for AVRCP common functions.
- Added _ct prefixes to all CT callback functions and feature APIs.
- Updated shell demo code for AVRCP API changes.

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi committed:
Bluetooth: AVRCP: Implement unit message reception on the AVRCP target

- Add a new callback to support the situation when a unit info command is
  received, and add an API to respond to the unit info command.

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi committed:
Bluetooth: Shell: AVRCP: support unit info request and response

Add avrcp_unit_info_req callback function in shell and a command to set
the unit info response.

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi commented:
@gzh-terry Can you please view the PR for the new proposal , thanks.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -419,20 +441,82 @@ static void avrcp_pass_through_handler(struct bt_avrcp *avrcp, uint8_t tid, stru
+static void avrcp_tg_unit_info_handler(struct bt_avrcp *avrcp, uint8_t tid, struct net_buf *buf,
+				       bt_avctp_cr_t cr)
+{
+	struct bt_avrcp_header *avrcp_hdr;
```
Whether do the fields `ctype`, `subunit_type` and `subunit_ID` need to be checked for a command frame?

makeshi replied:
Yes, currently just log dbg to printf these fields value, will update to add  fields ctype, subunit_type and subunit_ID checking for a command.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -780,5 +914,60 @@ int bt_avrcp_ct_register_cb(const struct bt_avrcp_ct_cb *cb)
 
 	avrcp_ct_cb = cb;
+
+	while (i < CONFIG_BT_MAX_CONN) {
+		atomic_set_bit(bt_avrcp_ct_pool[i++].state, BT_AVRCP_CT_ENABLE);
+	}
```
Can we simply it by checking if `avrcp_ct_cb != NULL`?

makeshi replied:
It looks could simply checking if avrcp_ct_cb != NULL, later if these is really need a state for AVRCP profile enable not, then I can add it back.

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -137,14 +143,16 @@ typedef enum __packed {
 
 struct bt_avrcp_unit_info_rsp {
-	uint8_t unit_type;
+	bt_avrcp_subunit_type_t unit_type;
```
The AV/C subunit type, also used for unit type ,

 
<img width="479" alt="unit_type" src="https://github.com/user-attachments/assets/3d4975dd-721d-4fae-9c37-b908ff67595e" />

PR also added it doxgen in the bt_avrcp_subunit_type_t 
/** @brief AV/C subunit type, also used for unit type */

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -457,10 +541,18 @@ static int avrcp_recv(struct bt_avctp *session, struct net_buf *buf)
-			handler[i].func(avrcp, tid, buf, cr);
+	ARRAY_FOR_EACH(tg_handler, i) {
+		if (avrcp_hdr->opcode == tg_handler[i].opcode) {
+			tg_handler[i].func(avrcp, tid, buf, cr);
```
Yes, the parameter cr is NOT needed, I will a new structure a avrcp_handler for tg, then can remove the cr parameter

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -513,16 +535,23 @@ struct bt_avrcp *bt_avrcp_connect(struct bt_conn *conn)
 	int err;
+	struct bt_avrcp *avrcp;
+
+	avrcp = get_new_connection(conn);
```
Will update to the function name avrcp_get_connection, then it does make sense also in disconnection operation

makeshi replied:
Will update to the function name avrcp_get_connection, then it does make sense also in disconnection operation

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -684,11 +786,14 @@ int bt_avrcp_ct_get_cap(struct bt_avrcp_ct *ct, uint8_t tid, uint8_t cap_id)
 	struct bt_avrcp_avc_pdu *pdu;
-
-        if ((ct == NULL) || (ct->avrcp == NULL)) {
+	if ((ct == NULL) || (ct->avrcp == NULL)) {
```
updated.

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -457,11 +542,19 @@ static int avrcp_recv(struct bt_avctp *session, struct net_buf *buf)
+				return 0;
+			}
 		}
 	}
```
updated.

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -684,11 +786,14 @@ int bt_avrcp_ct_get_cap(struct bt_avrcp_ct *ct, uint8_t tid, uint8_t cap_id)
 
+	if (avrcp_ct_cb == NULL) {
+		return -ENOTSUP;
+	}
```
Good suggestion, updated

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -419,20 +440,84 @@ static void avrcp_pass_through_handler(struct bt_avrcp *avrcp, uint8_t tid, stru
+/* ToDo */
+}
+
+static const struct avrcp_tg_handler tg_handler[] = {
```
code updated

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -514,6 +608,14 @@ int bt_avrcp_init(void)
+
+		i++;
+	}
+
```
updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -174,15 +235,20 @@ static int cmd_connect(const struct shell *sh, int32_t argc, char *argv[])
-		default_avrcp = NULL;
+	if ((default_ct != NULL) || (default_tg != NULL)) {
+		bt_avrcp_disconnect(default_conn);
+		default_ct = NULL;
+		default_tg = NULL;
```
Whether the flag should be cleared in disconnect callback?

makeshi replied:
Yes, flags need be cleared in disconnect callback, will update it.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -193,38 +209,69 @@ static struct bt_sdp_attribute avrcp_ct_attrs[] = {
+		LOG_ERR("Invalid parameter");
+		return NULL;
+	}
+	return &bt_avrcp_ct_pool[bt_conn_index(avrcp->acl_conn)];
```
There is an issue reported by the static code checking. It is better to add a `assert()` before using return value of `bt_conn_index(avrcp->acl_conn)`.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -193,38 +209,69 @@ static struct bt_sdp_attribute avrcp_ct_attrs[] = {
+		LOG_ERR("Invalid parameter");
+		return NULL;
+	}
+	return &bt_avrcp_tg_pool[bt_conn_index(avrcp->acl_conn)];
```
Also checking `bt_conn_index(avrcp->acl_conn)` here.

makeshi replied:
updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -193,38 +209,69 @@ static struct bt_sdp_attribute avrcp_ct_attrs[] = {
 
-	if ((avrcp_cb != NULL) && (avrcp_cb->connected != NULL)) {
-		avrcp_cb->connected(avrcp);
+	avrcp->acl_conn = avrcp->session.br_chan.chan.conn;
```
I think the `avrcp->acl_conn` should be set when creating connection or accepting connection request.

makeshi replied:
updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -492,37 +599,54 @@ int bt_avrcp_init(void)
-		return NULL;
+		return -ENOTCONN;
 	}
+	(void)memset(avrcp, 0, sizeof(struct bt_avrcp));
```
Same issue here.

makeshi replied:
updated

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -456,10 +562,11 @@ static int avrcp_accept(struct bt_conn *conn, struct bt_avctp **session)
 	if (!avrcp) {
 		return -ENOMEM;
 	}
+	(void)memset(avrcp, 0, sizeof(struct bt_avrcp));
```
updated.

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -222,9 +273,16 @@ static void avrcp_disconnected(struct bt_avctp *session)
+		avrcp_tg_cb->disconnected(get_avrcp_tg(avrcp));
 	}
+
+	bt_conn_unref(avrcp->acl_conn);
```
updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -513,16 +650,24 @@ struct bt_avrcp *bt_avrcp_connect(struct bt_conn *conn)
+	struct bt_avrcp *avrcp;
+
+	avrcp = avrcp_get_connection(conn);
+	if (!avrcp) {
```
The types on both sides of the equation need to be consistent. Suggest change `!avrcp` to `avrcp == NULL`.

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -192,31 +258,53 @@ static int cmd_disconnect(const struct shell *sh, int32_t argc, char *argv[])
+	if (default_tg != NULL) {
+		err = bt_avrcp_tg_send_unit_info_rsp(default_tg, tg_tid, &rsp);
+		if (!err) {
+			bt_shell_print("AVRCP send unit info response");
```
```suggestion
			shell_print(sh, "AVRCP send unit info response");
```

makeshi replied:
updated.

makeshi commented:
@gzh-terry Could you please help review the code change again? Thanks

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -301,107 +361,157 @@ static void avrcp_vendor_dependent_handler(struct bt_avrcp *avrcp, uint8_t tid,
-			rsp.extended_subunit_id = rsp.extended_subunit_type + rsp.max_subunit_id;
-			avrcp_cb->subunit_info_rsp(avrcp, tid, &rsp);
+	if ((avrcp_ct_cb != NULL) && (avrcp_ct_cb->subunit_info_rsp != NULL)) {
+		if (buf->len < 5) {
```
Use `BT_AVRCP_UNIT_INFO_CMD_SIZE` as well?

makeshi replied:
Yes, we can define new MACRO for 5 hard code, will update it

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -668,31 +821,60 @@ int bt_avrcp_get_cap(struct bt_avrcp *avrcp, uint8_t tid, uint8_t cap_id)
+{
+	struct net_buf *buf;
+
+	buf = avrcp_create_unit_pdu(avrcp, 0x0, BT_AVCTP_RESPONSE, BT_AVRCP_RSP_REJECTED);
```
Better to have a `tid` here to indicate which command is rejected.

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -33,17 +33,33 @@ LOG_MODULE_REGISTER(bt_avrcp);
 };
 
 #define AVRCP_AVCTP(_avctp) CONTAINER_OF(_avctp, struct bt_avrcp, session)
+#define IS_CT_ROLE_UNSUPPORTED() (avrcp_ct_cb == NULL)
```
We use calling bt_avrcp_ct_register_cb or bt_avrcp_tg_register_cb to determine which role is supported.
If the CT/TG corresponding callback is registered, then the CT or TG role is considered be supported. So add the two MACROs to code readable.
Or changing the macro to reflect this positive logic is more reasonable:

#define IS_CT_ROLE_SUPPORTED() (avrcp_ct_cb != NULL)
#define IS_TG_ROLE_SUPPORTED() (avrcp_tg_cb != NULL)

Or we could add comments to these two macros to avoid potential misunderstandings.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -33,22 +33,39 @@ LOG_MODULE_REGISTER(bt_avrcp);
 static struct bt_avrcp avrcp_connection[CONFIG_BT_MAX_CONN];
 static struct bt_avrcp_ct bt_avrcp_ct_pool[CONFIG_BT_MAX_CONN];
+static struct bt_avrcp_tg bt_avrcp_tg_pool[CONFIG_BT_MAX_CONN];
+static int bt_avrcp_send_unit_info_err_rsp(struct bt_avrcp *avrcp, uint8_t tid);
```
I think we can move the function to the front of function `avrcp_unit_info_cmd_handler()` to a void the function declaration.

makeshi replied:
This can be done, but it will involve moving quite a few functions, which will make the PR look like it has a lot of changes, firstly I have tried . Or we can consider doing it after this PR is merged

makeshi replied:
OK, will use a new commit to do this change to avoid PR look like it has a lot of changes.

makeshi committed:
Bluetooth: AVRCP: Adjust positions of internal static functions

- Move the static function locations to beginning of the file to the
  file to avoid function declarations.

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi commented:
@jhedberg @gzh-terry Could you please help review the code change again? Thanks

gzh-terry reviewed (approved):
LGTM

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -377,20 +471,39 @@ static int avrcp_recv(struct bt_avctp *session, struct net_buf *buf)
+			LOG_WRN("An AVRCP request is already ongoing.");
+			return -EALREADY;
+		}
+		if (avrcp_hdr->opcode == BT_AVRCP_OPC_PASS_THROUGH) {
```
Updated in the latest code.
